### PR TITLE
[SpecDecode][Kernel] Use Flashinfer for Rejection Sampling in Speculative Decoding

### PR DESCRIPTION
**Pull request by LiuXiaoxuanPKU**
End to end Speculative Decoding Performance (request latency):
`Draft: LLama-160M, Target: Vicuna-7B, batch size=8, input_len=256, output_len=512`

Before this PR:
```
Avg latency: 5.9652480507269505 seconds
10% percentile latency: 5.729408229794354 seconds
25% percentile latency: 5.794497653492726 seconds
50% percentile latency: 5.954964595614001 seconds
75% percentile latency: 6.124162045423873 seconds
90% percentile latency: 6.1757167306263 seconds
99% percentile latency: 6.3235187567165125 seconds

```
After this PR:
```
Avg latency: 5.717350374627858 seconds
10% percentile latency: 5.423373702447861 seconds
25% percentile latency: 5.50113928236533 seconds
50% percentile latency: 5.768905333476141 seconds
75% percentile latency: 5.950261808000505 seconds
90% percentile latency: 5.9829305820167065 seconds
99% percentile latency: 5.992807335173711 seconds
```

### CODE REVIEW

File: vllm/model_executor/layers/spec_decode_base_sampler.py
Comment (cadedaniel) [inline vllm/model_executor/layers/spec_decode_base_sampler.py:39]: nit: preference to move this to own variable like `Device = Union[int, str]` to prevent bloat.

File: vllm/spec_decode/spec_decode_worker.py
Comment (cadedaniel) [inline vllm/spec_decode/spec_decode_worker.py:598]: nit: `Get probabilities of target model, including bonus tokens.`

File: vllm/model_executor/layers/spec_decode_base_sampler.py
Comment (cadedaniel) [inline vllm/model_executor/layers/spec_decode_base_sampler.py:44]: nit: move this independent logic to own helper function

File: vllm/model_executor/layers/rejection_sampler.py
Comment (cadedaniel) [inline vllm/model_executor/layers/rejection_sampler.py:59]: should update the variable name `target_probs_with_bonus_probs ` here too https://github.com/vllm-project/vllm/blob/3979f69dbd8b36aa0286d09ac839c9cdc43a7f14/vllm/model_executor/layers/spec_decode_base_sampler.py#L215-L245

File: vllm/model_executor/layers/rejection_sampler.py
Comment (cadedaniel) [inline vllm/model_executor/layers/rejection_sampler.py:12]: nit: can add a comment explaining the motivation for a separate kernel? e.g. a blurb about how flash infer is faster

File: vllm/model_executor/layers/rejection_sampler.py
Comment (cadedaniel) [inline vllm/model_executor/layers/rejection_sampler.py:13]: wdyt about logging which one is loaded? makes debugging easier

File: vllm/model_executor/layers/rejection_sampler.py
Comment (cadedaniel) [inline vllm/model_executor/layers/rejection_sampler.py:152]: Can you you add a docstring?

File: vllm/model_executor/layers/rejection_sampler.py
Comment (cadedaniel) [inline vllm/model_executor/layers/rejection_sampler.py:158]: nit: suggest returning immediately, so readers do not have to consider this branch when reading the rest of the function

File: vllm/model_executor/layers/rejection_sampler.py
Comment (cadedaniel) [inline vllm/model_executor/layers/rejection_sampler.py:90]: nit: use `chain_speculative_sampling is not None` to be precise (None equality is different than [Falsey](https://stackoverflow.com/questions/39983695/what-is-truthy-and-falsy-how-is-it-different-from-true-and-false))

File: vllm/model_executor/layers/rejection_sampler.py
Comment (cadedaniel) [inline vllm/model_executor/layers/rejection_sampler.py:85]: nit: the `raise_if_incorrect_input` function should check that the input is correct, before any modification. otherwise it is less helpful as it becomes more like an assertion on internal state, not on correct input

File: vllm/model_executor/layers/rejection_sampler.py
Comment (cadedaniel) [reply vllm/model_executor/layers/rejection_sampler.py:90]: suggest also adding a comment on this branch for readability

**Review by cadedaniel — COMMENTED** [review]
LGTM. two qs:
* can we run the correctness test for both paths? specifically the convergence test (all of the e2e tests depend on this for temp>0)
* can we make sure there is no perf regression for the non-FlashInfer path?

**Comment by comaniac** [comment]
Note: there's a bugfix for correctness issue in this sampling kernel (https://github.com/flashinfer-ai/flashinfer/pull/425), so we may want to bump FlashInfer to the next release.

**Comment by yzh119** [comment]
> Note: there's a bugfix for correctness issue in this sampling kernel ([flashinfer-ai/flashinfer#425](https://github.com/flashinfer-ai/flashinfer/pull/425)), so we may want to bump FlashInfer to the next release.

Yes @LiuXiaoxuanPKU 's number are measured with flashinfer main branch where #425 was already merged. This PR depends on flashinfer v0.1.4.

File: vllm/model_executor/layers/spec_decode_base_sampler.py
Comment (LiuXiaoxuanPKU) [reply vllm/model_executor/layers/spec_decode_base_sampler.py]: will leave it as is for now. People can refactor later if they reuse this part of code.

**Comment by LiuXiaoxuanPKU** [comment]
@cadedaniel 
The PR is ready for a second review. It should be pass speculative decoding tests, but I do have some questions & concerns.

1. Currently, we update the metrics in the [_create_output](https://github.com/vllm-project/vllm/blob/7025b11d949b4efeb2584690c35f919c77027368/vllm/model_executor/layers/spec_decode_base_sampler.py#L125), however, with the rejection sampling kernel, we will not go through that code pass. Therefore, we will update the metrics outside (https://github.com/vllm-project/vllm/blob/7439daf60fa820185851aebf606bc411264f2068/vllm/model_executor/layers/rejection_sampler.py#L121), not sure if it's good practice.
2. The rejection sampler does not return vllm speculative decoding's definition of `accepted` tokens, therefore the [metric](https://github.com/vllm-project/vllm/blob/7439daf60fa820185851aebf606bc411264f2068/vllm/model_executor/layers/rejection_sampler.py#L121) is not meaningful for now. I update in that way to just pass the CI.
3. In the current CI, we install flashinfer by default. Therefore, all speculative decoding tests (including rejection sampling) are using the kernel. Do we want to keep the test for the old code? Will that be a bit expensive for CI?

**Comment by LiuXiaoxuanPKU** [comment]
@cadedaniel  Updates:
1. modify the flashinfer kernel to return the metric required by vllm spec dec metric. PR: https://github.com/flashinfer-ai/flashinfer/pull/450
2. change rejection sampler to only mainly test the flashinfer backend.
3. add tests to compare the flashinfer and nonflashinfer backend results.

This PR is ready, CI tests might fail because we will need flashinfer to release and add the latest flashinfer to CI. Tests passed locally.

File: tests/samplers/test_rejection_sampler.py
Comment (comaniac) [inline tests/samplers/test_rejection_sampler.py:85]: I feel you can leave this test untouched, and just rename the follow test to "test_flashinfer_backed".

File: tests/samplers/test_rejection_sampler.py
Comment (comaniac) [inline tests/samplers/test_rejection_sampler.py:135]: Just add use_flashinfer to the parameter so that we could test both cases?

File: vllm/model_executor/layers/rejection_sampler.py
Comment (comaniac) [inline vllm/model_executor/layers/rejection_sampler.py:20]: Can we control this kernel using VLLM_USE_FLASHINFER_SAMPLER as well?

File: vllm/model_executor/layers/rejection_sampler.py
Comment (comaniac) [inline vllm/model_executor/layers/rejection_sampler.py:50]: Could be just warning?

File: vllm/model_executor/layers/spec_decode_base_sampler.py
Comment (comaniac) [inline vllm/model_executor/layers/spec_decode_base_sampler.py:8]: We already have vllm.utils.Device. Better to use another name or just not define it.

**Review by comaniac — COMMENTED** [review]
Overall LGTM. Just nits.

**Comment by cadedaniel** [comment]
Will review today.

File: vllm/model_executor/layers/rejection_sampler.py
Comment (LiuXiaoxuanPKU) [reply vllm/model_executor/layers/rejection_sampler.py:50]: Ideally, when disable_bonus_tokens, the bonus token should be -1.
However, if we use flashinfer and set disable_bonus_tokens, the bonus token will still have values (!= -1), which makes the results incorrect. I guess it might be better to just fail here?

File: vllm/model_executor/layers/rejection_sampler.py
Comment (cadedaniel) [inline vllm/model_executor/layers/rejection_sampler.py:21]: nit: also log in the case when flashinfer is not used (again, makes debugging easier)

File: vllm/model_executor/layers/rejection_sampler.py
Comment (cadedaniel) [inline vllm/model_executor/layers/rejection_sampler.py:35]: update docstring

File: vllm/model_executor/layers/rejection_sampler.py
Comment (cadedaniel) [reply vllm/model_executor/layers/rejection_sampler.py:50]: we can remove the `disable_bonus_token` path completely now that https://github.com/vllm-project/vllm/issues/4212 is fixed.

but if it's too much work let's just leave it as assert, that way "no failure" means user gets the experience we planned for them instead of missing a warning and getting subpar perf

File: vllm/model_executor/layers/rejection_sampler.py
Comment (cadedaniel) [inline vllm/model_executor/layers/rejection_sampler.py:107]: nit: add comment for when this case is expected

File: vllm/model_executor/layers/rejection_sampler.py
Comment (cadedaniel) [inline vllm/model_executor/layers/rejection_sampler.py:112]: code readability: it is good practice to reduce the number of conditions available in the code, both so it's easier to read and so there are less bugs due to less possible permutations of state.

in this case, suggest having a single condition `self.use_flashinfer`, which is validated in the constructor (e.g. if use_flashinfer is True, then assert chain_speculative_sampling is not None). this will simplify the code and reduce the amount of state.

File: vllm/model_executor/layers/rejection_sampler.py
Comment (cadedaniel) [inline vllm/model_executor/layers/rejection_sampler.py:125]: shouldn't this be `(output_token_ids != -1).sum()` to include the bonus token? seems adding `batch_size` causes incorrect calculation, correct me if I'm wrong

File: vllm/model_executor/layers/rejection_sampler.py
Comment (cadedaniel) [inline vllm/model_executor/layers/rejection_sampler.py:215]: I love the simplicity of the normal case 😍

File: vllm/model_executor/layers/spec_decode_base_sampler.py
Comment (cadedaniel) [reply vllm/model_executor/layers/spec_decode_base_sampler.py:8]: I said earlier to move it to a separate one to make the code have less boilerplate 😆 . will defer to your best judgement @LiuXiaoxuanPKU .

File: tests/samplers/test_rejection_sampler.py
Comment (cadedaniel) [inline tests/samplers/test_rejection_sampler.py:208]: Can you add a docstring explaining the test methodology? e.g. as a maintainer, I want to know what value this test provides, how it tests what it tests

File: tests/samplers/test_rejection_sampler.py
Comment (cadedaniel) [inline tests/samplers/test_rejection_sampler.py:346]: can we run this test both with and without flashinfer? without it, we won't have confidence the code is correct.

File: tests/samplers/test_rejection_sampler.py
Comment (cadedaniel) [inline tests/samplers/test_rejection_sampler.py:101]: we should keep this assertion in; otherwise the `disable_bonus_tokens` path is not tested. it's fine to delete that path but we should delete it if that's what we want.

File: tests/samplers/test_rejection_sampler.py
Comment (cadedaniel) [reply tests/samplers/test_rejection_sampler.py:346]: sorry, I think I may have miscommunicated offline what we need here:
* it is OK for the e2e tests to run with either flashinfer or non-flashinfer. we do not have to run all the tests twice.
* for the rejection sampler tests, we _need_ to run all tests for both paths. this is because otherwise we are not confident that the rejection sampler is correct.

if we don't satisfy these two conditions, then we can have a scenario where the E2E runs do not enable flashinfer for whatever reason, and our normal rejection sampling code isn't well tested, causing us to have correctness issues that are not caught until after merge (messy and sad!).

File: vllm/model_executor/layers/rejection_sampler.py
Comment (LiuXiaoxuanPKU) [reply vllm/model_executor/layers/rejection_sampler.py:125]: The `emitted_token_num` in flashinfer is the number of emitted tokens, stops at the first token that violates the condition `p >= q' and does not include recovery/bonus token. Therefore, we need to add `batch_size` here.

File: tests/samplers/test_rejection_sampler.py
Comment (LiuXiaoxuanPKU) [reply tests/samplers/test_rejection_sampler.py:101]: Will add it back. We need another PR to clean up the `disable_bonus_tokens` field.

File: vllm/model_executor/layers/rejection_sampler.py
Comment (cadedaniel) [reply vllm/model_executor/layers/rejection_sampler.py:135]: got it. can you add a comment here for readability?

**Comment by yzh119** [comment]
FYI: flashinfer v0.1.6 wheels are ready: https://github.com/flashinfer-ai/flashinfer/releases/tag/v0.1.6

File: tests/samplers/test_rejection_sampler.py
Comment (cadedaniel) [inline tests/samplers/test_rejection_sampler.py:225]: I think this needs to go out of the helper function, else the rand will be different

File: tests/samplers/test_rejection_sampler.py
Comment (cadedaniel) [inline tests/samplers/test_rejection_sampler.py:231]: nit: docstring goes on top

File: tests/samplers/test_rejection_sampler.py
Comment (cadedaniel) [reply tests/samplers/test_rejection_sampler.py:225]: I realize there's an error here -- it should be `torch.rand(...) <= 0.5`

File: vllm/model_executor/layers/rejection_sampler.py
Comment (yzh119) [inline vllm/model_executor/layers/rejection_sampler.py:139]: Why not just align flashinfer's behavior and this API's?

File: tests/samplers/test_rejection_sampler.py
Comment (LiuXiaoxuanPKU) [reply tests/samplers/test_rejection_sampler.py:225]: I think I will just remove the rand, we should just fix the generator for each request in the batch instead of fixing it with 50% probability.

**Review by cadedaniel — APPROVED** [review]

**Pull request merged**
